add GPIO support on i.MX 943 A-Core

JiafeiPan opened the pull request:
This PR is to add GPIO support for i.MX 943 EVK A-Core and with some minimal board document fix.

JiafeiPan committed:
dts: nxp_mimx943_a55: add GPIO device nodes

Added all GPIO device nodes in i.MX 943 Cortex-A Core SoC dts.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
tests: gpio_basic_api: add imx943_evk_mimx94398_a55 overly

Add board overlay for i.MX943 EVK Cortex-A Core.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
drivers: gpio: rgpio: use default pad config value for SCMI platform

If the platform uses SCMI pinctrl driver, pinctrl regitster can't accessed
by CPU Core directly, and currently SCMI pinctrl driver has no API to read
back the register value, so use default pad config value for GPIO pad
configuration, and in theory we could use a fixed pad config value in this
driver as each new GPIO configuration has no relation with previous
configuration.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
tests: gpio: gpio_basic_api: config to input if disconnected not supported

If GPIO_DISCONNECTED is not supported by GPIO driver, then try to configure
the pin to be input, otherwise there will be some unstable signal between
two test pins.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
drivers: gpio: rgpio: not support GPIO_DISCONNECTED

The hardware don't support GPIO_DISCONNECTED.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
boards: imx943_evk: fix document format issue

Fixed the document format issue.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
boards: imx943_evk: add GPIO in supported list for A55

Add GPIO in board yaml's supported list

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

yangbolu1991 commented on drivers/gpio/gpio_mcux_rgpio.c:
```diff
@@ -21,6 +21,9 @@
 	((const struct mcux_rgpio_config *)(_dev)->config)
 #define DEV_DATA(_dev) ((struct mcux_rgpio_data *)(_dev)->data)
 
+/* Default PAD config value for SCMI clock: Pull down, Slight Fast Slew Rate, X4 driver strength */
```
`SCMI clock`?

JiafeiPan replied:
It should be SCMI pinctrl, has been fixed, thanks.

yangbolu1991 commented:
May I know the reason to use default pad config value? Better to add in commit message.

[drivers: gpio: rgpio: use default pad config value for SCMI platform](https://github.com/zephyrproject-rtos/zephyr/pull/89878/commits/762426f523eee1039f38aa8e0588af166302d005)

Thanks.

JiafeiPan commented on drivers/gpio/gpio_mcux_rgpio.c:
```diff
@@ -68,10 +75,15 @@ static int mcux_rgpio_configure(const struct device *dev,
 	}
 
+#if defined(CONFIG_PINCTRL_IMX_SCMI)
+	/* For SCMI Pinctrl platform, set default PAD config value for SCMI clock. */
```
It should be SCMI pinctrl, has been fixed, thanks.

JiafeiPan commented:
> May I know the reason to use default pad config value? Better to add in commit message.
> 
> [drivers: gpio: rgpio: use default pad config value for SCMI platform](https://github.com/zephyrproject-rtos/zephyr/pull/89878/commits/762426f523eee1039f38aa8e0588af166302d005)
> 
> Thanks.

   If the platform uses SCMI pinctrl driver, pinctrl regitster can't accessed by
    CPU Core directly, and currently SCMI pinctrl driver has no API to read back
    the register value, so use default pad config value for GPIO pad configuration,
    and in theory we could use a fixed pad config value in this driver as each
    new GPIO configuration has no relation with previous configuration.

Updated commit message.

PetervdPerk-NXP commented on dts/arm64/nxp/nxp_mimx943_a55.dtsi:
```diff
@@ -192,6 +192,85 @@
+		gpio-controller;
+		#gpio-cells = <2>;
+		ngpios = <28>;
+		gpio-reserved-ranges = <10 6>;
```
I see you use the `gpio-reserved-ranges` feature. However I don't see it implemented in GPIO ISR handler.
These reserved pins can still trigger an interrupt, I would filter these already out in the ISR otherwise an assertion can hit.

JiafeiPan replied:
Hi, @PetervdPerk-NXP,  could you please help me to understand why reverved pins can still trigger an interrupt? in RGPIO driver, "mcux_rgpio_pin_interrupt_configure" is used to configure interrupt for each pin, and in this function it has already use port_pin_mask filter reserved pins, that means reverved pins ICR(Interrupt Control Register)'s IRQC will be keeped as reset value "0000": ISF is disabled, so that it ISF bit for reserved pins will keep as 0 and "base->ISFR[0] = int_flags" in mcux_rgpio_port_isr() will not have the interrupt active flags for reserved pins. so in a word, we don't worry about reverved pins can trigger interrupt in current device driver.

PetervdPerk-NXP replied:
Let's say the M33 and A55 core both run a Zephyr kernel.

Buth both allocate the GPIO1.
M33 use does `gpio-reserved-ranges = <0 16>;`
A55 use does `gpio-reserved-ranges = <16 16>;`

If the A55 for example enables an IRQ on GPIO1 Pin 1 then it will write the ICR register setting Pin. When GPIO1 peripheral triggers an interrupt, then both A55 and M33 will receive the interrupt, but the M33 will forward the interrupt as well to zephyr kernel even though it's reserved.

JiafeiPan committed:
drivers: gpio: rgpio: only handle usable pin's interrupt

If gpio-reserved-ranges to reserve some pins which used by other CPU
Core's OS, we could only handle usable pins owned by current CPU
Core in interrupt handler.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
